src/libYARP_dev/include/yarp/dev/IJoypadController.h

```diff
@@ -155,6 +171,97 @@ class  YARP_dev_API yarp::dev::IJoypadController
     */
     virtual bool getTouch(unsigned int touch_id, yarp::sig::Vector& value) = 0;
 };
+
+class YARP_dev_API yarp::dev::IJoypadEvent
+{
+public:
+    virtual ~IJoypadEvent();
+
+    template <typename T> struct joyData
+    {
+        unsigned int m_id;
+        T            m_datum;
+
+        joyData(unsigned int id, const T& datum)
+        {
+            m_id    = id;
+            m_datum = datum;
+        }
+    };
+
+    virtual void action(std::vector<joyData<float> >             buttons,
+                        std::vector<joyData<double> >            axes,
+                        std::vector<joyData<unsigned char> >     hats,
+                        std::vector<joyData<yarp::sig::Vector> > trackBalls,
+                        std::vector<joyData<yarp::sig::Vector> > sticks,
+                        std::vector<joyData<yarp::sig::Vector> > Touch) = 0;
+};
+
+
+
+class YARP_dev_API yarp::dev::IJoypadEventDriven : yarp::os::RateThread,
+                                                   public yarp::dev::IJoypadController
+{
+private:
+    yarp::dev::IJoypadEvent*       m_event;
+    bool                           EventDrivenEnabled;
+    std::vector<float>             old_buttons;
+    std::vector<double>            old_axes;
+    std::vector<unsigned char>     old_hats;
+    std::vector<yarp::sig::Vector> old_trackballs;
+    std::vector<yarp::sig::Vector> old_sticks;
+    std::vector<yarp::sig::Vector> old_touches;
+protected:
+    virtual bool getRawAxisCount(unsigned int& axis_count) = 0;
+    virtual bool getRawButtonCount(unsigned int& button_count) = 0;
+    virtual bool getRawTrackballCount(unsigned int& Trackball_count) = 0;
+    virtual bool getRawHatCount(unsigned int& Hat_count) = 0;
+    virtual bool getRawTouchSurfaceCount(unsigned int& touch_count) = 0;
+    virtual bool getRawStickCount(unsigned int& stick_count) = 0;
+    virtual bool getRawStickDoF(unsigned int stick_id, unsigned int& DoF) = 0;
+    virtual bool getRawButton(unsigned int button_id, float& value) = 0;
+    virtual bool getRawTrackball(unsigned int trackball_id, yarp::sig::Vector& value) = 0;
+    virtual bool getRawHat(unsigned int hat_id, unsigned char& value) = 0;
+    virtual bool getRawAxis(unsigned int axis_id, double& value) = 0;
+    virtual bool getRawStick(unsigned int stick_id, yarp::sig::Vector& value, JoypadCtrl_coordinateMode coordinate_mode) = 0;
+    virtual bool getRawTouch(unsigned int touch_id, yarp::sig::Vector& value) = 0;
+    using IJoypadController::m_actions;
+    using IJoypadController::executeAction;
+    using IJoypadController::parseActions;
+
+public:
+
+    virtual bool getAxisCount(unsigned int& axis_count) YARP_OVERRIDE YARP_FINAL;
```

Comment:
I thought so too.. but it seems not. compiler gives you the Wsuggest-override warning (with only final) and you can put final in a virtual method also in the first declaration (the one in the top class).. it doesn't make any sense to me but is possible..
so you have to specify final and override..